src/ipips/ipip-0412.md

```diff
@@ -0,0 +1,240 @@
+a useful order.
+
+The existing Trustless Gateway specification does not provide a mechanism for
+negotiating the order of blocks in CAR responses.
+
+This IPIP aims to improve the status quo.
+
+## Detailed design
+
+CAR content type
+([`application/vnd.ipld.car`](https://www.iana.org/assignments/media-types/application/vnd.ipld.car))
+already supports `version` parameter, which allows gateway to indicate which
+CAR flavour is returned with the response.
+
+The proposed solution introduces two new parameters for the content type headers
+in HTTP requests and responses: `order` and `dups`.
+
+The `order` parameter allows the client to indicate its preference for a
+specific block order in the CAR response, and the `dups` parameter specifies
+whether duplicate blocks are allowed in the response.
+
+### Signaling in Request
+
+Content type negotiation is based on section 12.5.1 of :cite[rfc9110].
+
+Clients MAY indicate their preferred block order by sending an `Accept` header in
+the HTTP request. The `Accept` header format is as follows:
+
+```
+Accept: application/vnd.ipld.car; version=1; order=dfs; dups=y
+```
+
+In the future, when more orders or parameters exist, clients will be able to
+specify a list of preferences, for example:
+
+```
+Accept: application/vnd.ipld.car;order=foo, application/vnd.ipld.car;order=dfs;dups=y;q=0.5
+```
+
+The above example is a list of preferences, the client would really like to use
+the hypothetical `order=foo` however if this isn't available it would accept
+`order=dfs` with `dups=y` instead (lower priority indicated via `q` parameter,
+as noted in :cite[rfc9110]).
+
+#### `order` CAR content type parameter
+
+The `order` parameter accepts the following values:
+
+- `dfs`: [Depth-First Search](https://en.wikipedia.org/wiki/Depth-first_search)
+  order, allows for streaming responses with minimal memory usage
+- `rnd`: Unknown (random) order, the implicit default when `order` parameter is missing.
+
+#### `dups` CAR content type parameter
+
+The `dups` parameter specifies whether duplicate blocks (the same block
+occuring multiple times in the requested DAG) will be present in the CAR
+response.
+
+It accepts two values:
+- `y`: duplicate blocks are allowed
+- `n`: duplicates are not allowed
+
+When allowed (`y`), light clients are able to  discard blocks after
+reading them, removing the need for caching in-memory or on-disk.
+
+<!-- TODO: do we need a  parameter for inclusion of identity CIDs?
+It seems to be only relevant in Filecoin due to legacy hiccup:
+https://github.com/ipfs/specs/pull/330#issuecomment-1274106892 -->
+
+### Signaling in Response
+
+The Trustless Gateway MUST always respond with a `Content-Type` header that includes
+information about all supported/known parameters, even if the client did not
+specify them in the request.
+
+The `Content-Type` header format is as follows:
+
+```
+Content-Type: application/vnd.ipld.car;version=1;order=dfs;dups=y
```

Comment:
@hannahhoward  to unblock work in service worker, biforst-gateway/boxo and gateway-conformance before this IPIP is ratified, is this acceptable default to implement in Lassie / Saturn L1?  

Your implementation is free to be more strict and only support one set of parameters.
Namely, the ask here would be for Lassie/Saturn L1 to return blocks in Depth-First Search order & sending duplicates.

Reply:
@lidel you guys need duplicates?

Reply:
we specifically DO NOT send duplicates for the time being, it's non-trivial to throw them back in.

Reply:
cc: @aschmahmann 

Basically, current Lassie is:
Content-Type: application/vnd.ipld.car;version=1;order=dfs;dups=n

For "dups=y" and maybe this is not clear from the spec, are dups simply allowed, or required? (that is worth putting in the spec in fact)

Reply:
IMO the point of duplicates is to enable people to write clients that stream results and do not need to keep blocks buffered.  If people prefer a different name/nomenclature that's fine, but the end result of `Content-Type: application/vnd.ipld.car;version=1;order=dfs;dups=y` should be that you are doing a preorder DFS traversal where even if the same block is encountered multiple times in the traversal it is still emitted once per time it is needed as part of the traversal.

For example, consider having go-ipld-prime be hooked up to a traversal and an inbound car such that:
1. If you ask it for a block other than the next block it has available it errors
2. It can throw away blocks after the first time they're used and the traversal will still complete

This should be usable to verifiably stream out all of the file bytes of a UnixFS file with only a single network request and a small fixed amount of memory (e.g. a few MiBs).

Reply:
> it's non-trivial to throw them back in.

This is somewhat off-topic/implementation specific so if there's another place you'd prefer to engage lmk, but why is this the case?

IIUC this should be fairly straight-forward to implement given that go-ipld-prime traversals do not buffer data in memory and the strategy you take towards populating your CAR files is to do a traversal and on block-load get the block. If the go-car functions are forcing deduplication on you and not giving you an alternative way out that seems like a fixable problem.

Reply:
Yes I agree it should be non-trivial.  For reasons, ATM, it's not.

Do you need `dups=y` on your side for Rhea? I can talk to @rvagg if so.

Reply:
https://github.com/filecoin-project/lassie/issues/239#issuecomment-1550736498

Would be good to know if we need to support this for the *->saturn->lassie flow or just in the lassie->provider flow. As I said in the above comment I think it'd be useful option to have for data coming out of lassie for streaming cases (to avoid [this](https://github.com/ipld/go-car/blob/92d28eb23ba496288e309dc9672edf13cf9329c8/cmd/car/extract.go#L378-L444)) but there'd be some work to do to make it happen, hopefully mainly on the testing side.

But, we should factor in the dangers of allowing this, we open up that space where deeply interlinked graphs can explode both in traversal recursion and, with this option, output data. You could probably make a retriever turn a few Mb of source data into Gb. Nice attack amplification potential with that.

Reply:
> For "dups=y" and maybe this is not clear from the spec, are dups simply allowed, or required?

@hannahhoward  updated this section to clarify that value returned in `dups` means required behavior (MUST).
(goal of this IPIP is to introduce opt-in signaling mechanism about how deterministic the block order in the response is, so the client can apply optimizations)

> Would be good to know if we need to support this for the *->saturn->lassie flow or just in the lassie->provider flow. 

@rvagg  imo Saturn needs it on the  client→saturn→lassie flow  because we want to have more and more light clients[^1] to hit  Saturn L1 directly, and `dups=y` will make it possible to process response cheaper/faster by removing the extra storage steps similar to the one you linked. 


[^1]: current example is  Saturn's own [browser clients](https://github.com/filecoin-saturn/browser-client), but in a few months this could include [native IPFS in Chromium based on the Trustless Gateway spec](https://github.com/ipfs/ipfs-blog/pull/560) in places like Brave, Electron etc.


> You could probably make a retriever turn a few Mb of source data into Gb. Nice attack amplification potential with that.

@rvagg  I need some help unpacking this. My understanding is that it is already possible to craft a huge blob from well-available data, and then request it from any gateway to waste resources, does not even need `?format=car`.

The difference with `dups=y` is that a retrieval client like Lassie would not pay the cost of retrieving all the blocks, 
and instead return a single one multiple times. 

The end result is the same (wasted resources, CPU/bandwidth capped) but Lassie did not waste resources of other peers to retrieve the payload.

 Is there a version of this where more resources are wasted than it is already possible?